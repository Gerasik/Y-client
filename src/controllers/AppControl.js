import AppModel from '../models/AppModel';
import AppMain from '../views/AppMain/AppMain';
import AppSlider from '../views/AppSlider/AppSlider';
import AppClips from '../views/AppClips/AppClips';

export default class App {
  constructor() {
    this.parametrs = {
      apiKey: 'AIzaSyAboqwlqPy-2bcz8E7RPz9mTmlhoQVCbvU',
      urlSearch: null,
      nextPageToken: null,
    };
    this.slider = null;
    this.currSlide = null;
    this.cursorPosition = null;
  }

  start() {
    const main = new AppMain(App.search.bind(this));
    main.render();
  }

  static async search(ev) {
    ev.preventDefault();
    if (this.parametrs.urlSearch) {
      App.slider.element.remove();
      document.querySelector('.pagination').remove();
    }
    const search = document.querySelector('.header-search__search');

    this.parametrs.urlSearch = search.value;

    const model = new AppModel(this.parametrs);
    const data = await model.getClipNames();
    App.slider = new AppSlider(App.move, App.lock);
    App.slider.render();
    App.currSlide = 0;
    App.parametrs = this.parametrs;
    App.parametrs.nextPageToken = data.nextPageToken;
    App.renderClips(await data.data);
  }

  static renderClips(data) {
    let clipsCount;
    if (window.outerWidth > 1300) {
      clipsCount = 4;
    } else if (window.outerWidth <= 1300 && window.outerWidth > 768) {
      clipsCount = 3;
    } else if (window.outerWidth <= 768 && window.outerWidth > 425) {
      clipsCount = 2;
    } else if (window.outerWidth <= 425) {
      clipsCount = 1;
    }
    data.forEach((element) => {
      if (this.slider.element.lastElementChild.childElementCount === clipsCount) {
        this.slider.addSlide();
      }
      const clips = new AppClips(element, this.slider.element);
      clips.render();
    });
  }

  static unify(e) { return e.changedTouches ? e.changedTouches[0] : e; }

  static lock(e) { App.cursorPosition = App.unify(e).clientX; }

  static async move(e) {
    const pagination = document.querySelector('.pagination');
    if (e.type === 'click') {
      switch (e.target.parentNode.classList[0]) {
        case 'pagination__prev':
        case 'pagination__curr':
          if (App.currSlide > 0) App.currSlide -= 1;
          break;
        case 'pagination__next-page':
        case 'pagination__next':
          App.currSlide += 1;
          break;
        default:
      }
      pagination.childNodes[1].firstChild.innerHTML = App.currSlide + 1;
      pagination.childNodes[2].firstChild.innerHTML = App.currSlide + 2;
      if (App.currSlide > -1) {
        App.slider.element.childNodes.forEach(a => a.style.setProperty('transform', `translateX(-${App.currSlide}00%)`));
        if (App.currSlide === App.slider.element.childElementCount - 2) {
          const model = new AppModel(App.parametrs);
          const data = await model.getClipNames();
          App.parametrs.nextPageToken = data.nextPageToken;
          App.renderClips(await data.data);
        }
      }
    }
    if (App.cursorPosition || App.cursorPosition === 0) {
      const dx = App.unify(e).clientX - App.cursorPosition;
      if (dx > 0 && App.currSlide > 0) {
        App.currSlide -= 1;
      } else if (dx < 0) {
        App.currSlide += 1;
      }
      pagination.childNodes[1].firstChild.innerHTML = App.currSlide + 1;
      pagination.childNodes[2].firstChild.innerHTML = App.currSlide + 2;
      if (App.currSlide > -1) {
        App.slider.element.childNodes.forEach(a => a.style.setProperty('transform', `translateX(-${App.currSlide}00%)`));
        if (App.currSlide === App.slider.element.childElementCount - 2) {
          const model = new AppModel(App.parametrs);
          const data = await model.getClipNames();
          App.parametrs.nextPageToken = data.nextPageToken;
          App.renderClips(await data.data);
        }
      }
      App.cursorPosition = null;
    }
  }
}
