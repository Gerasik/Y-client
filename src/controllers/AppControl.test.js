import AppControl from './AppControl';

describe('AppControl.prototype.render', () => {
  it('Should be an instance of Function', () => {
    expect(AppControl.prototype.start).toBeInstanceOf(Function);
  });

  it('Should be render search form', () => {
    AppControl.prototype.start.call();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('Should be render clips', () => {
    const context = {
      slider: {
        element: document.body,
        addSlide: document.body.appendChild(document.createElement('div')),
      },
    };
    const data = [{
      title: 'Title1',
      description: 'Desc1',
      image: './img1.png',
      date: 'publishedAt1',
      autor: 'title1',
      views: 11,
      id: 1,
    }, {
      title: 'Title2',
      description: 'Desc2',
      image: './img2.png',
      date: 'publishedAt2',
      autor: 'title2',
      views: 22,
      id: 2,
    }, {
      title: 'Title3',
      description: 'Desc3',
      image: './img3.png',
      date: 'publishedAt3',
      autor: 'title3',
      views: 33,
      id: 3,
    }];
    AppControl.renderClips.call(context, data);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
