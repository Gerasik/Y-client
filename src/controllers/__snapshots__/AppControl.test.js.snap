// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppControl.prototype.render Should be render clips 1`] = `
"<header class=\\"header-search\\"><form><input type=\\"submit\\" value=\\"Search\\"><input type=\\"text\\" placeholder=\\"Search\\" class=\\"header-search__search\\"></form></header><div><div class=\\"clip\\"><a href=\\"https://www.youtube.com/watch?v=1\\" class=\\"clip__title\\">Title1</a>
                      <img class=\\"clip__img\\" src=\\"undefined\\" alt=\\"Title1\\" width=\\"undefined\\" height=\\"undefined\\">
                      <span class=\\"clip__autor\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" width=\\"20px\\"><path d=\\"M256 288c79.5 0 144-64.5 144-144S335.5 0 256 0 112 64.5 112 144s64.5 144 144 144zm128 32h-55.1c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16H128C57.3 320 0 377.3 0 448v16c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48v-16c0-70.7-57.3-128-128-128z\\"></path></svg>
                      title1</span>
                      <span class=\\"clip__date\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\" width=\\"20px\\"><path d=\\"M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z\\"></path></svg>
                      publishedA</span>
                      <span class=\\"clip__views\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 576 512\\" width=\\"20px\\"><path d=\\"M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z\\"></path></svg>
                      11</span>
                      <p class=\\"clip__description\\">Desc1</p></div><div class=\\"clip\\"><a href=\\"https://www.youtube.com/watch?v=2\\" class=\\"clip__title\\">Title2</a>
                      <img class=\\"clip__img\\" src=\\"undefined\\" alt=\\"Title2\\" width=\\"undefined\\" height=\\"undefined\\">
                      <span class=\\"clip__autor\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" width=\\"20px\\"><path d=\\"M256 288c79.5 0 144-64.5 144-144S335.5 0 256 0 112 64.5 112 144s64.5 144 144 144zm128 32h-55.1c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16H128C57.3 320 0 377.3 0 448v16c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48v-16c0-70.7-57.3-128-128-128z\\"></path></svg>
                      title2</span>
                      <span class=\\"clip__date\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\" width=\\"20px\\"><path d=\\"M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z\\"></path></svg>
                      publishedA</span>
                      <span class=\\"clip__views\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 576 512\\" width=\\"20px\\"><path d=\\"M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z\\"></path></svg>
                      22</span>
                      <p class=\\"clip__description\\">Desc2</p></div><div class=\\"clip\\"><a href=\\"https://www.youtube.com/watch?v=3\\" class=\\"clip__title\\">Title3</a>
                      <img class=\\"clip__img\\" src=\\"undefined\\" alt=\\"Title3\\" width=\\"undefined\\" height=\\"undefined\\">
                      <span class=\\"clip__autor\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" width=\\"20px\\"><path d=\\"M256 288c79.5 0 144-64.5 144-144S335.5 0 256 0 112 64.5 112 144s64.5 144 144 144zm128 32h-55.1c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16H128C57.3 320 0 377.3 0 448v16c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48v-16c0-70.7-57.3-128-128-128z\\"></path></svg>
                      title3</span>
                      <span class=\\"clip__date\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\" width=\\"20px\\"><path d=\\"M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z\\"></path></svg>
                      publishedA</span>
                      <span class=\\"clip__views\\">
                      <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 576 512\\" width=\\"20px\\"><path d=\\"M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z\\"></path></svg>
                      33</span>
                      <p class=\\"clip__description\\">Desc3</p></div></div>"
`;

exports[`AppControl.prototype.render Should be render search form 1`] = `"<header class=\\"header-search\\"><form><input type=\\"submit\\" value=\\"Search\\"><input type=\\"text\\" placeholder=\\"Search\\" class=\\"header-search__search\\"></form></header>"`;
